### PR TITLE
Locking support

### PR DESCRIPTION
**Pull request by AlSchlo**
This PR introduced the locking functionality and slightly reorganizes the general CryptoCookies infrastructure.

### CODE REVIEW

File: backend/contracts/cc/CryptoCookies.sol [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+/// Copyright © 2023 Chainlabs Switzerland SA
+/// All Rights Reserved
+
+
+pragma solidity ^0.8.18;
+
+/// @title CryptoCookies
+/// @author Chainlabs Switzerland SA
+/// @notice 
+contract CryptoCookies { // Upgradable? isDAO (heritance)
```
Comment (emduc) [inline backend/contracts/cc/CryptoCookies.sol:11]: Not sure of the difference between this and CCCore

File: backend/contracts/cc/locking/CCLocking.sol [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	/// @notice Locks a given amount of TOKEN to obtain an equivalent amount of
+	/// yield and capital tokens. May only be called before the end of the period.
+	/// Requires a prior ERC20 token approval.
+	/// @param _amount The amount to lock.
+	function lock(uint256 _amount) external {
+		require(PERIOD_END < block.timestamp);
+		require(LOCKED_TOKEN.transferFrom(msg.sender, address(this), _amount));
+
+		CAPITAL_TOKEN.mint(msg.sender, _amount);
+		YIELD_TOKEN.mint(msg.sender, _amount);
```
Comment (emduc) [inline backend/contracts/cc/locking/CCLocking.sol:68]: Shouldn't it then be staked automatically?

File: backend/contracts/cc/locking/ERC20Mintable.sol [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+/// @title An ERC20 Mintable Contract
+/// @author Chainlabs Switzerland SA
+contract ERC20Mintable is ERC20, AccessControl, ERC20Permit {
+	/// @notice Constructs a mintable ERC20 token contract.
+	/// Caller has the sole right to assign new minters through the admin role.
+	constructor(
+		string memory _name,
+		string memory _symbol
+	) ERC20(_name, _symbol) ERC20Permit(_name) {
+		_setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
```
Comment (emduc) [inline backend/contracts/cc/locking/ERC20Mintable.sol:20]: how does that work such that the admin can be the locking contract?

File: backend/contracts/wtf/distributors/FdgDistr.sol [hunk 4]
```diff
@@ -43,15 +41,6 @@ contract FdgDistr is BaseERC20Distr {
...
 
-	/// @notice Changes the stake of a given app in the distributor.
-	/// Can only be called by Wtf's gauge contract.
-	/// @param _app The target app.
-	/// @param _change The change in stake.
-	function appChangeStake(address _app, int256 _change) external {
-		require(msg.sender == GAUGE);
-		_userChangeStake(_app, _change);
-	}
-
```
Comment (emduc) [inline backend/contracts/wtf/distributors/FdgDistr.sol:54]: how is it done then

File: backend/package-lock.json [hunk 5]
```diff
@@ -828,40 +828,6 @@
         "rlp": "^2.2.3"
       }
     },
-    "node_modules/@morgan-stanley/ts-mocking-bird": {
```
Comment (emduc) [inline backend/package-lock.json:831]: this file is automatically generated and should not be comitted right?

File: backend/contracts/cc/CryptoCookies.sol (hunk 1 above)
Comment (AlSchlo) [reply backend/contracts/cc/CryptoCookies.sol:11]: CCCore is mostly the core distribution functionality of CryptoCookies. It is directly connected to the Fudge Finance ecosystem. The CryptoCookies contract is the central point of access: it deploys the DAO, the Core, the competition environment and other surrounding features. Most probably CryptoCookies will inherit CCCore.

File: backend/contracts/cc/locking/CCLocking.sol (hunk 2 above)
Comment (AlSchlo) [reply backend/contracts/cc/locking/CCLocking.sol:68]: Great suggestion, I'll add a function that does both to bundle both transactions into one.

File: backend/contracts/cc/locking/ERC20Mintable.sol (hunk 3 above)
Comment (AlSchlo) [reply backend/contracts/cc/locking/ERC20Mintable.sol:20]: The locking contract will deploy the ERC20Mintable and thus be the msg.sender.

File: backend/contracts/wtf/distributors/FdgDistr.sol (hunk 4 above)
Comment (AlSchlo) [reply backend/contracts/wtf/distributors/FdgDistr.sol:54]: Changed the parent contract and added a default function userChangeStake().

File: backend/package-lock.json (hunk 5 above)
Comment (AlSchlo) [reply backend/package-lock.json:831]: You're right!

**Pull request merged**